packages/angular/ssr/src/routes/route-tree.ts

```diff
@@ -0,0 +1,224 @@
+   * @param metadata - The metadata associated with the route.
+   */
+  insert(route: string, metadata: Omit<RouteTreeNodeMetadata, 'route'>): void {
+    let node = this.root;
+    const normalizedRoute = stripTrailingSlash(route);
+    const segments = normalizedRoute.split('/');
+
+    for (const segment of segments) {
+      // Parameterized route segment, replace it with a wildcard
+      const normalizedSegment = segment[0] === ':' ? '*' : segment;
+      let childNode = node.children?.get(normalizedSegment);
+
+      if (!childNode) {
+        childNode = {
+          segment: normalizedSegment,
+          insertionIndex: -1,
+        };
+        node.children ??= new Map();
+        node.children.set(normalizedSegment, childNode);
+      }
+
+      node = childNode;
+    }
+
+    // Store the full route at the leaf node
+    node.metadata = {
+      route: normalizedRoute,
+      ...metadata,
+    };
+    node.insertionIndex = this.insertionIndexCounter++;
+  }
+
+  /**
+   * Matches a route against the route tree, returning the best match
+   * based on the lowest insertion index.
+   * @param route - The route to match.
+   * @returns The best matching route or null if no match is found.
+   */
+  match(route: string): RouteTreeNodeMetadata | null {
+    let bestMatchMetadata: RouteTreeNodeMetadata | null = null;
+    let lowestIndex = Infinity;
+
+    /**
+     * Recursively traverses the route tree to find the best match.
+     */
+    const traverse = (currentNode: RouteTreeNode, remainingSegments: string[]): void => {
+      const { insertionIndex, metadata, children } = currentNode;
+      if (!remainingSegments.length) {
+        // Reached a leaf node, check if it's a better match
+        if (metadata && insertionIndex < lowestIndex) {
+          bestMatchMetadata = metadata;
+          lowestIndex = insertionIndex;
+        }
+
+        return;
+      }
+
+      if (!children) {
+        return;
+      }
+
+      const [segment, ...restSegments] = remainingSegments;
+
+      // Try to match the exact segment
+      let childNode = children.get(segment);
+      if (childNode) {
+        traverse(childNode, restSegments);
+      }
+
+      // Try to match with a wildcard
+      childNode = children.get('*');
+      if (childNode) {
+        traverse(childNode, restSegments);
+      }
+    };
+
+    const segments = stripTrailingSlash(route).split('/');
+    traverse(this.root, segments);
+
+    return bestMatchMetadata;
+  }
+
+  /**
+   * Converts the route tree into a nested object representation.
+   * @returns A nested object where each key is a route segment and each value is an object containing
+   *          metadata and possibly further nested route segments.
+   */
+  toObject(): RouteNodeObjectRepresentation {
+    const result: RouteNodeObjectRepresentation = {};
+
+    /**
+     * Recursively traverses the route tree to build the nested object.
+     */
+    const traverse = (
```

Comment:
Consider: Make a generic `traverse` or `walk` function which just returns a generator which iterates over all the nodes in the tree. Then `toObject` can use that to iterate and focus on building the object representation.

This would allow better decoupling between the two problems of 1) walking the `RouteTree` and 2) building the object representation.

Reply:
Good call, updated.